src/operations/blas3/gemm_partial_local.hpp

```diff
@@ -0,0 +1,465 @@
+     * enabled */
+    if (double_buffer) {
+      extract_input_blocks(local_id, 0, scratch_ptr, rhs_scratch_ptr,
+                           global_m_offset, global_n_offset, global_k_offset,
+                           check_m_limit, check_n_limit);
+    }
+
+    index_t tile_id = 0;
+    /* Loop over all tiles allocated to this particular workgroup size */
+    do {
+      id.barrier(cl::sycl::access::fence_space::local_space);
+
+      // Start loading the next tile
+      index_t next_tile = double_buffer ? (tile_id + 1) : tile_id;
+      const bool tile_nb_check = do_check<double_buffer>(next_tile < num_tiles);
+      if (tile_nb_check) {
+        extract_input_blocks(local_id, next_tile, scratch_ptr, rhs_scratch_ptr,
+                             global_m_offset, global_n_offset, global_k_offset,
+                             check_m_limit, check_n_limit);
+      }
+
+      if (!double_buffer) {
+        id.barrier(cl::sycl::access::fence_space::local_space);
+      }
+
+      // Calculate offsets into the temporary memory.
+      index_t lhs_offset =
+          (double_buffer * (tile_id & 1) * lhs_tile_mem_size) + m_local_id;
+      index_t rhs_offset =
+          (double_buffer * (tile_id & 1) * rhs_tile_mem_size) + n_local_id;
+
+      /* Loop over the values of a single tile */
+      for (index_t k = 0; k < tile_size_dim_k; k++) {
+        auto idx = 0;
+        auto rhs_index = 0;
+#pragma unroll
+        for (index_t wLPTN = 0; wLPTN < work_per_thread_n; wLPTN++) {
+          // load a RHS element from the scratch buffer
+          const element_t privateRhs = rhs_scratch_ptr[rhs_index + rhs_offset];
+
+          index_t lhs_index = 0;
+#pragma unroll
+          for (index_t wLPTM = 0; wLPTM < work_per_thread_m; wLPTM++) {
+            // load a LHS element from the scratch buffer
+            const element_t privateLhs = scratch_ptr[lhs_index + lhs_offset];
+
+            // Perform a manual MAD.
+            private_res[wLPTM + idx] += (privateLhs * privateRhs);
+
+            lhs_index += local_thread_size_m;
+          }
+          idx += work_per_thread_m;
+          rhs_index += local_thread_size_n;
+        }
+        lhs_offset += ld_lhs_tile;
+        rhs_offset += ld_rhs_tile;
+      }
+      tile_id++;
+    } while (tile_id < num_tiles);
+
+    // Store the final results in the cube buffer
+    index_t slice_col = (ngroup_id * tile_size_dim_n) + (n_local_id);
+    const index_t slice_row_offset =
+        (mgroup_id * tile_size_dim_m) + (m_local_id);
+    const index_t cube_depth_offset = kgroup_id * m_ * n_;
+    index_t cube_index = slice_col * m_;
+    index_t private_index_offset = 0;
+
+#pragma unroll
+    for (index_t wLPTN = 0; wLPTN < work_per_thread_n; wLPTN++) {
+      index_t private_index = private_index_offset;
+
+      index_t slice_row = slice_row_offset;
+#pragma unroll
+      for (index_t wLPTM = 0; wLPTM < work_per_thread_m; wLPTM++) {
+        if (slice_row < m_ && slice_col < n_) {
```

Comment:
in every write, you have to perform the check which reduces the performance, the same rule that you use for computing the internal block will apply here for the store as well. So none of the internal blocks should go through the check and only the edges should go there. so the store function should be wrap in a method and based on whether or not it is internal you need to go through a true or false path

Reply:
This change will go with the other optimizations that I included in #178 (unless we want to move these optimizations here, but I feel like we should merge this MR as soon as possible and add optimizations in another MR)